controllers/cassandradatacenter_controller.go

```diff
@@ -0,0 +1,288 @@
+	lastNodeStart := rc.Datacenter.Status.LastServerNodeStarted
+	cooldownTime := time.Until(lastNodeStart.Add(twentySecs))
+
+	if cooldownTime > 0 {
+		logger.Info("Ending reconciliation early because a server node was recently started")
+		secs := time.Duration(1 + int(cooldownTime.Seconds()))
+		return ctrl.Result{RequeueAfter: secs * time.Second}, nil
+	}
+
+	if rc.Datacenter.Status.QuietPeriod.After(time.Now()) {
+		logger.Info("Ending reconciliation early because the datacenter is in a quiet period")
+		cooldownTime = rc.Datacenter.Status.QuietPeriod.Sub(time.Now())
+		secs := time.Duration(1 + int(cooldownTime.Seconds()))
+		return ctrl.Result{RequeueAfter: secs * time.Second}, nil
+	}
+
+	res, err := rc.CalculateReconciliationActions()
+	if err != nil {
+		logger.Error(err, "calculateReconciliationActions returned an error")
+		rc.Recorder.Eventf(rc.Datacenter, "Warning", "ReconcileFailed", err.Error())
+	}
+	return res, err
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CassandraDatacenterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	log := r.Log.
+		WithName("cassandradatacenter_controller")
+
+	// Create a new managed controller builder
+	c := ctrl.NewControllerManagedBy(mgr).
+		Named("cassandradatacenter-controller").
+		WithLogger(log).
+		For(&api.CassandraDatacenter{}, builder.WithPredicates(predicate.GenerationChangedPredicate{}))
+
+	// Watch for changes to primary resource CassandraDatacenter
+	// c = c.Watches(
+	// 	&source.Kind{Type: &api.CassandraDatacenter{}},
+	// 	&handler.EnqueueRequestForObject{},
+	// 	// This allows us to update the status on every reconcile call without
+	// 	// triggering an infinite loop.
+	// 	builder.WithPredicates(predicate.GenerationChangedPredicate{}))
+
+	// Here we list all the types that we create that are owned by the primary resource.
+	//
+	// Watch for changes to secondary resources StatefulSets, PodDisruptionBudgets, and Services and requeue the
+	// CassandraDatacenter that owns them.
+	managedByCassandraOperatorPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return oplabels.HasManagedByCassandraOperatorLabel(e.Object.GetLabels())
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return oplabels.HasManagedByCassandraOperatorLabel(e.Object.GetLabels())
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return oplabels.HasManagedByCassandraOperatorLabel(e.ObjectOld.GetLabels()) ||
+				oplabels.HasManagedByCassandraOperatorLabel(e.ObjectNew.GetLabels())
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return oplabels.HasManagedByCassandraOperatorLabel(e.Object.GetLabels())
+		},
+	}
+
+	// NOTE: We do not currently watch PVC resources, but if we did, we'd have to
+	// account for the fact that they might use the old managed-by label value
+	// (oplabels.ManagedByLabelDefunctValue) for CassandraDatacenters originally
+	// created in version 1.1.0 or earlier.
+
+	c = c.Watches(
```

Comment:
Can we replace each of the `Watches` with the `Owns` method of `Builder` where we are using a `handler.EnqueueRequestForOwner`?

Reply:
I had it that way, but now I can't remember why I had to rewrite them back to watches..

Reply:
Okay, so I think I did the change to make all the other types consistent. But I can replace some of those Watches (not all) with Owns.